engine/src/main/coffee/engine/core/world/export.coffee

```diff
@@ -0,0 +1,260 @@
+
+{ difference, find, isEmpty, toObject } = require('brazierjs/array')
+{ id, tee }                             = require('brazierjs/function')
+{ fold }                                = require('brazierjs/maybe')
+
+NLType = require('../typechecker')
+
+# Yo!  This file expects that basically all of its functions will be called in the context
+# of the `World` object.  That is, they should be called within methods on `World`, using
+# `<function>.call(this)`. --JAB (12/10/17)
+
+# (String|(Number, Number, Number)|(Number, Number, Number, Number)) => ExportedColor
+exportColor = (color) ->
+  if NLType(color).isNumber()
+    new ExportedColorNum(color)
+  else if NLType(color).isList()
+    [r, g, b, a] = color
+    if a?
+      new ExportedRGBA(r, g, b, a)
+    else
+      new ExportedRGB(r, g, b)
+  else
+    throw new Error("Unrecognized color format: #{JSON.stringify(color)}")
+
+# (String) => BreedReference
+exportBreedReference = (breedName) ->
+  new BreedReference(breedName.toLowerCase())
+
+# (Patch) => PatchReference
+exportPatchReference = (patch) ->
+  new PatchReference(patch.pxcor, patch.pycor)
+
+# (Turtle) => TurtleReference
+exportTurtleReference = (turtle) ->
+  breed = new BreedNamePair(turtle.getBreedNameSingular(), turtle.getBreedName().toLowerCase())
+  new TurtleReference(breed, turtle.id)
+
+# (Link) => LinkReference
+exportLinkReference = (link) ->
+  breed = new BreedNamePair(link.getBreedNameSingular(), link.getBreedName().toLowerCase())
+  new LinkReference(breed, link.end1.id, link.end2.id)
+
+# (Agent) => AgentReference
+exportAgentReference = (agent) ->
+  type = NLType(agent)
+  if type.isNobody() or agent.isDead()
+    NobodyReference
+  else if type.isLink()
+    exportLinkReference(agent)
+  else if type.isPatch()
+    exportPatchReference(agent)
+  else if type.isTurtle()
+    exportTurtleReference(agent)
+  else
+    throw new Error("Cannot make agent reference out of: #{JSON.stringify(agent)}")
+
+# (Agent) => (String) => Any
+exportWildcardVar = (agent) -> (varName) ->
+
+  exportWildcardValue = (value) ->
+    type = NLType(value)
+    if type.isAgent() or type.isNobody()
+      exportAgentReference(value)
+    else if value.getSpecialName?()?
+      new BreedReference(value.getSpecialName().toLowerCase())
+    else if type.isLinkSet()
+      new ExportedLinkSet(value.toArray().map(exportLinkReference))
+    else if type.isPatchSet()
+      new ExportedPatchSet(value.toArray().map(exportPatchReference))
+    else if type.isTurtleSet()
+      new ExportedTurtleSet(value.toArray().map(exportTurtleReference))
+    else if type.isCommandLambda()
+      new ExportedCommandLambda(value.nlogoBody)
+    else if type.isReporterLambda()
+      new ExportedReporterLambda(value.nlogoBody)
+    else if type.isList()
+      value.map(exportWildcardValue)
+    else
+      value
+
+  exportWildcardValue(agent.getVariable(varName))
+
+# () => Object[Any]
+exportMetadata = ->
+  # TODO: IMPLEMENT FILENAME
```

Comment:
You may want to provide the link to the NetLogo/NetLogo issue that's holding you up here.